crates/modelardb_compression_python/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+/* Copyr ight 2023 The ModelarDB Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Implementation of a Python interface for ModelarDB's compression library.
+
+mod convert;
+
+use std::sync::Arc;
+
+use arrow::array::{Array, ArrayRef, BinaryArray, Float32Array, UInt64Array, UInt8Array};
+use arrow::datatypes::Schema;
+use arrow::error::ArrowError;
+use arrow::record_batch::RecordBatch;
+use modelardb_common::schemas::{COMPRESSED_SCHEMA, QUERY_SCHEMA, UNCOMPRESSED_SCHEMA};
+use modelardb_common::types::{TimestampArray, TimestampBuilder, ValueArray, ValueBuilder};
+use modelardb_compression::{self, models};
+use pyo3::import_exception;
+use pyo3::prelude::{pyfunction, pymodule, wrap_pyfunction, PyAny, PyModule, PyResult, Python};
+
+import_exception!(pyarrow, ArrowException);
+
+/// Initialize the module by registering functions in the Python module.
+#[pymodule]
+fn modelardb_compression_python(_python: Python<'_>, py_module: &PyModule) -> PyResult<()> {
+    py_module.add_function(wrap_pyfunction!(compress, py_module)?)?;
+    py_module.add_function(wrap_pyfunction!(decompress, py_module)?)?;
+
+    Ok(())
+}
+
+/// Compress a [`RecordBatch`] containing an univariate time series within `error_bound`. An
+/// `univariate_id` can optionally be set to specify which univariate time series the compressed
+/// segments was produced from, if none is set zero is used. An [`ArrowException`] is returned if
+/// the schema of `uncompressed` does not match [`UNCOMPRESSED_SCHEMA`] or if [`pyo3`] cannot
+/// convert the arguments from Python to Rust and the result from Rust to Python.
+#[pyfunction(univariate_id = 0)]
+fn compress<'a>(
+    uncompressed: &'a PyAny,
+    error_bound: &'a PyAny,
+    univariate_id: u64,
+    python: Python<'a>,
+) -> PyResult<&'a PyAny> {
+    // Convert the record batch from Python to Rust and validate the arguments.
+    let uncompressed = convert::python_record_batch_to_rust(uncompressed)?;
+    if !fields_has_equal_types(&uncompressed.schema(), &UNCOMPRESSED_SCHEMA.0) {
+        return Err(ArrowException::new_err(
+            "Uncompressed can only contain millisecond timestamps and float32 values.".to_owned(),
```

Comment:
Maybe change to `Uncompressed data can...` instead.

Reply:
See this [comment](https://github.com/ModelarData/ModelarDB-RS/pull/68#discussion_r1102579970).